### ISSUE TITLE
Use issuer claim for selecting which openidConnectClient to use for a JWT inbound propagation request

### ISSUE THREAD

**Issue by teddyjtorres** [OpenLiberty/open-liberty#20168]
## Description

Use issuer claim for selecting which openidConnectClient to use for a JWT inbound propagation request.
AS-IS: Complicated auth filters are need to select which openidConnectClient configuration to use.
TO-BE: Allow using the issuer claim from the JWT access token to select the openidConnectClient with a matching issuer configuration. More than one matching openidConnectClient will result in a error.

For a JWT with an issuer claim value of `"https://hostname/op1"` and a server.xml file configured with,

```
<openidConnectClient id="RP1" issuer="https://hostname/op1"  ... />
<openidConnectClient id="RP2" issuer="https://hostname/op2" authFilterRef="rp2filter" .../>

<authFilter id="rp2filter">
</authFilter>
```

the RP1 configuration is selected as long as the "rp2filter" does not apply to the request. Therefore, an attempt to find which openidConnectClient applies to the request will be done using the authFilters first, using the following algorithm

1. Run auth filter.  If auth filter matches, choose that config.
2. Check issuer from the JWT against issuer of all config elements.  If the issuer from the JWT matches a single config element, choose that config.
3. If the issuer from the JWT matches the issuer of multiple config elements, choose between all of those config elements in a non-deterministic manner.
4. If the issuer does not match the issuer of any config elements, choose between all of the config elements in a non-deterministic manner.

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## Documents

When available, add links to required feature documents. Use "N/A" to mark particular documents which are not required by the feature.

- Aha: Externally raised RFE ([Aha](https://cloud-platform.ideas.aha.io/))
  - Link the RFE with this issue
- UFO: Link to Upcoming Feature Overview
  - Set the Box link to be publicly accessible, with a long expiration (10 years)
    - click "Share" > select "People with link" > click "Link Settings" > under "Link Experation" select "Disable Shared Link on" > set an experation date ~10 years into the future
    - If you lack permissions, contact [OpenLiberty/release-architect](https://github.com/orgs/OpenLiberty/teams/release-architect)
- FTS: https://github.com/OpenLiberty/open-liberty/issues/20761
- Beta Blog: Link to Beta Blog Post GH Issue
- GA Blog: https://github.com/OpenLiberty/open-liberty/issues/21783

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## Process Overview

- [Prioritization](#prioritization)
- [Design](#design)
- [Implementation](#implementation)
- [Legal and Translation](#legal-and-translation)
- [Beta](#beta)
- [GA](#ga)
  - [Focal Point Approvals](#focal-point-approvals-complete-by-feature-complete-date)
- [Other Deliverables](#other-deliverables)

### General Instructions

The process steps occur roughly in the order as presented. Process steps occasionally overlap.

Each process step has a number of tasks which must be completed or must be marked as not applicable ("N/A").

Unless otherwise indicated, the tasks are the responsibility of the Feature Owner or a Delegate of the Feature Owner.

If you need assistance, reach out to the [OpenLiberty/release-architect](https://github.com/orgs/OpenLiberty/teams/release-architect).

**Important: Labels are used to trigger particular steps and must be added as indicated.** 

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## **Prioritization** (Complete Before Development Starts)

The ([OpenLiberty/chief-architect](https://github.com/orgs/OpenLiberty/teams/chief-architect)) and area leads are responsible for prioritizing the features and determining which features are being actively worked on.

### **Prioritization**
- [x] Feature added to the "New" column of the [Open Liberty project board](https://github.com/orgs/OpenLiberty/projects/2)
  - Epics can be added to the board in one of two ways:
    - From this issue, use the "Projects" section to select the appropriate project board.
    - From the appropriate project board click "Add card" and select your Feature Epic issue
- [x] Priority assigned
  - Attend the Liberty Backlog Prioritization meeting

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## **Design** (Complete Before Development Starts)

Design preliminaries determine whether a formal design, which will be provided by an Upcoming Feature Overview (UFO) document, must be created and reviewed.  A formal design is required if the feature requires any of the following: UI, Serviceability, SVT, Performance testing, or non-trivial documentation/ID.

### **Design Preliminaries**
- [ ] UI requirements identified. (Owner and [UI focal point](https://github.com/orgs/OpenLiberty/teams/ui-approvers))
- [x] ID requirements identified. (Owner and [ID focal point](https://github.com/orgs/OpenLiberty/teams/id-approvers)) - https://github.com/OpenLiberty/docs/issues/5646
   - Refer to [Documenting Open Liberty](https://github.com/OpenLiberty/open-liberty/wiki/Documenting-Open-Liberty).
   - Feature Owner adds label `ID Required`, if non-trivial documentation needs to be created by the ID team.
   - ID adds label `ID Required - Trivial`, if no design will be performed and only trivial ID updates are needed.
- [ ] Serviceability Requirements Identified. (Owner and [Serviceability focal point](https://github.com/orgs/OpenLiberty/teams/serviceability-approvers))
- [ ] SVT Requirements Identified. (Owner and [SVT focal point](https://github.com/orgs/OpenLiberty/teams/svt-approvers))
- [ ] Performance testing requirements identified. (Owner and [Performance focal point](https://github.com/orgs/OpenLiberty/teams/performance-approvers))

### **Design**
- [ ] POC Design / UFO review requested.
  - Owner adds label `Design Review Request`
- [ ] POC Design / UFO review scheduled. (David Chang)
- [ ] POC Design / UFO review completed.
- [ ] POC / UFO Review follow-ons completed.
- [ ] Design / UFO approved. ([OpenLiberty/chief-architect](https://github.com/orgs/OpenLiberty/teams/chief-architect)) or N/A
  - ([OpenLiberty/chief-architect](https://github.com/orgs/OpenLiberty/teams/chief-architect)) adds label `Design Approved`
  - Add the public link to the UFO in Box to the [Documents](#documents) section.
  - The UFO must always accurately reflect the final implementation of the feature. Any changes must be first approved. Afterwards, update the UFO by creating a copy of the original approved slide(s) at the end of the deck and prepend "OLD" to the title(s). A single updated copy of the slide(s) should take the original's place, and have its title(s) prepended with "UPDATED".

### **No Design**
- [x] No Design requested.
  - Owner adds label `No Design Approval Request`
- [x] No Design / No UFO approved. ([OpenLiberty/chief-architect](https://github.com/orgs/OpenLiberty/teams/chief-architect)) or N/A
  - Approver adds label `No Design Approved`

### **FAT Documentation**
- [x] "Feature Test Summary" child task created
  - Use the [Feature Test Summary Template](https://github.com/OpenLiberty/open-liberty/issues/new?assignees=&labels=Feature+Test+Summary&template=feature_test_summary.md&title=)
  - Add FTS issue link to the [Documents](#documents) section.

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## **Implementation**

A feature must be [prioritized](https://github.com/orgs/OpenLiberty/projects/2) and socialized (or `No Design Approved`) before any implementation work may begin and is the minimum before any beta code may be delivered.  All new Liberty content must be inaccessible in our GA releases until it is [Feature Complete](#feature-complete) by either marking it `kind=noship` or [beta fencing](#beta-code) it.  
Code may not GA until this feature has obtained the "Design Approved" or "No Design Approved" label, along with all other tasks outlines in [GA](#ga) section.

### **Feature Development Begins**
- [x] Add the `In Progress` label

## **Legal and Translation**

In order to avoid last minute blockers and significant disruptions to the feature, the **legal items need to be done as early in the feature process as possible**, either in design or as early into the development as possible.  Similarly, translation is to be done concurrently with development.  Both **MUST** be completed before Beta or GA is requested.

### **Legal** (Complete before Feature Complete Date)
- [x] Changed or new open source libraries are cleared and approved, or N/A. (Legal Release Services/Cass Tucker/Release PM).
- [x] Licenses and Certificates of Originality (COOs) are updated, or N/A

### **Translation** (Complete 1 week before Feature Complete Date)
- [x] PII updates are merged, or N/A. Note timing with translation shipments.

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## **Beta**

In order to facilitate early feedback from users, all new features and functionality should first be released as part of a beta release.

### **Beta Code**
- [x] Beta fence the functionality
  - `kind=beta`, `ibm:beta`, `ProductInfo.getBetaEdition()`
- [ ] Beta development complete and feature ready for inclusion in a beta release
  - Add label `target:beta` and the appropriate `target:YY00X-beta` (where YY00X is the targeted beta version).
- [ ] Feature delivered into beta
  - ([OpenLiberty/release-manager](https://github.com/orgs/OpenLiberty/teams/release-manager)) adds label `release:YY00X-beta` (where YY00X is the first beta version that included the functionality).

### **Beta Blog** (Complete 1.5 weeks before beta eGA)
- [ ] Beta blog issue created and populated using the [Open Liberty BETA blog post](https://github.com/OpenLiberty/open-liberty/issues/new/choose) template.
  - Add a link to the beta blog issue in the [Documents](#documents) section.

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## **GA**

A feature is ready to GA after it is Feature Complete and has obtained all necessary Focal Point Approvals. 

### **Feature Complete**
- [x] Feature implementation and tests completed.
  - [x] All PRs are merged.
  - [x] All epic and child issues are closed.
  - [x] All stop ship issues are completed.
- [x] Legal: all necessary approvals granted.
- [x] Translation: All messages translated or sent for translation for upcoming release 
- [x] GA development complete and feature ready for inclusion in a GA release
  - Add label `target:ga` and the appropriate `target:YY00X` (where YY00X is the targeted GA version).
  - Inclusion in a release requires the completion of all Focal Point Approvals.

### **Focal Point Approvals** (Complete by Feature Complete Date)

These occur only after GA of this feature is requested (by adding a `target:ga` label).  GA of this feature may not occur until all approvals are obtained.

### **All Features**
- [x] **FAT** All Tests complete and running successfully in SOE or N/A. ([OpenLiberty/fat-approvers](https://github.com/orgs/OpenLiberty/teams/fat-approvers))
  - Approver adds label `focalApproved:fat`.
- [x] **Demo** Demo is scheduled for an upcoming EOI or N/A. ([OpenLiberty/demo-approvers](https://github.com/orgs/OpenLiberty/teams/demo-approvers))
  - Approver adds label `focalApproved:demo`.
- [x] **Globalization** Translation and TVT are complete or N/A. ([OpenLiberty/globalization-approvers](https://github.com/orgs/OpenLiberty/teams/globalization-approvers))
  - Approver adds label `focalApproved:globalization`.

### **Design Approved Features**
- [ ] **Accessibility** Accessibility testing completed or N/A. ([OpenLiberty/accessibility-approvers](https://github.com/orgs/OpenLiberty/teams/accessibility-approvers))
  - Approver adds label `focalApproved:accessibility`.
- [ ] **APIs/Externals** Externals have been reviewed or N/A.  ([OpenLiberty/externals-approvers](https://github.com/orgs/OpenLiberty/teams/externals-approvers))
  - Approver adds label `focalApproved:externals` 
- [ ] **ID** Documentation is complete or N/A. ([OpenLiberty/id-approvers](https://github.com/orgs/OpenLiberty/teams/id-approvers))
  - Approver adds label `focalApproved:id`.
  - > **_NOTE:_**  If only trivial documentation changes are required, you may reach out to the ID Feature Focal to request a `ID Required - Trivial` label.  Unlike features with regular ID requirement, those with `ID Required - Trivial` label do not have a hard requirement for a Design/UFO.
- [ ] **Performance** Performance testing is complete or N/A. ([OpenLiberty/performance-approvers](https://github.com/orgs/OpenLiberty/teams/performance-approvers))
  - Approver adds label `focalApproved:performance`.
- [ ] **Serviceability** Serviceability has been addressed or N/A. ([OpenLiberty/serviceability-approvers](https://github.com/orgs/OpenLiberty/teams/serviceability-approvers))
  - Approver adds label `focalApproved:sve`.
- [ ] **STE** Skills Transfer Education chart deck is complete or N/A. ([OpenLiberty/ste-approvers](https://github.com/orgs/OpenLiberty/teams/ste-approvers))
  - Approver adds label `focalApproved:ste`.
- [ ] **SVT** System Verification Test is complete or N/A. ([OpenLiberty/svt-approvers](https://github.com/orgs/OpenLiberty/teams/svt-approvers))
  - Approver adds label `focalApproved:svt`.

### **Remove Beta Fencing** (Complete by Feature Complete Date)
- [x] Beta guards are removed, or N/A
  - Only after all necessary Focal Point Approvals have been granted.

### **GA Blog** (Complete by Feature Complete Date)
- [x] GA Blog issue created and populated using the [Open Liberty GA release blog post](https://github.com/OpenLiberty/open-liberty/issues/new/choose) template.
  - Add a link to the GA Blog issue in the [Documents](#documents) section.

### **Post GA**
- [ ] Replace `target:YY00X` label with the appropriate `release:YY00X`. ([OpenLiberty/release-manager](https://github.com/orgs/OpenLiberty/teams/release-manager))

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## **Other Deliverables**

- [ ] **OL Guides** OL Guides assessment is complete or N/A. ([Yee-Kang Chang](https://github.com/yeekangc))
- [ ] **Standalone Feature Blog Post** A blog post specifically about your feature or N/A.  ([OpenLiberty/release-architect](https://github.com/orgs/OpenLiberty/teams/release-architect))
  - This should be strongly considered for larger or more prominent features.
  - Follow [instructions](https://github.com/OpenLiberty/blogs/tree/draft#writing-and-publishing-blog-posts-on-the-openlibertyio-blog) in the blogs repo.
- [ ] **WDT** Liberty Developer Tools work is complete or N/A. (Leonard Theivendra)

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

**Comment by barbj** [comment]
1. Run auth filter. If auth filter matches, choose that config.
    -  ok 
2. Check issuer from the JWT against issuer of all config elements. If the issuer from the JWT matches a single config element, choose that config.
    - ok
3. If the issuer from the JWT matches the issuer of multiple config elements, choose between all of those config elements in a non-deterministic manner.
   - how is this ok?  I think that you need to give the admin an option for which config to use.  If you don't do that, they could get random failures.  For instance, say that you have multiples and one has sigAlg=RS256 and another has 512.  All of your "extra" JWTs come in as 512.  If you randomly pick between the two clients, your "extra" JWTs are going to randomly fail.  If you give a config option, the admin can tag the 512 to use (or the 256 to not use? whatever)  There is a bevy of other reasons why a JWT would pass with one client and fail with another with the OP of the same issuer.  With authFilters, we say: if you don't like that, make your filter better.  We can't make anything "better" with the issuer.  The only way to make it better is to make the config better, which would be a property.  If you give a property and then they have then _issuer+property_ is all the same, then I can agree with the non-deterministic aspect.  I think that the (_USE-ME_!) option is needed here and wouldn't really be a burden on either the code or the admin.
4. If the issuer does not match the issuer of any config elements, choose between all of the config elements in a non-deterministic manner.
   - why is that not a 401?

**Comment by teddyjtorres** [comment]
> 1. Run auth filter. If auth filter matches, choose that config.
>    
>    * ok
> 2. Check issuer from the JWT against issuer of all config elements. If the issuer from the JWT matches a single config element, choose that config.
>    
>    * ok
> 3. If the issuer from the JWT matches the issuer of multiple config elements, choose between all of those config elements in a non-deterministic manner.
>    
>    * how is this ok?  I think that you need to give the admin an option for which config to use.  If you don't do that, they could get random failures.  For instance, say that you have multiples and one has sigAlg=RS256 and another has 512.  All of your "extra" JWTs come in as 512.  If you randomly pick between the two clients, your "extra" JWTs are going to randomly fail.  If you give a config option, the admin can tag the 512 to use (or the 256 to not use? whatever)  There is a bevy of other reasons why a JWT would pass with one client and fail with another with the OP of the same issuer.  With authFilters, we say: if you don't like that, make your filter better.  We can't make anything "better" with the issuer.  The only way to make it better is to make the config better, which would be a property.  If you give a property and then they have then _issuer+property_ is all the same, then I can agree with the non-deterministic aspect.  I think that the (_USE-ME_!) option is needed here and wouldn't really be a burden on either the code or the admin.
> 4. If the issuer does not match the issuer of any config elements, choose between all of the config elements in a non-deterministic manner.
>    
>    * why is that not a 401?

Hi.  Choosing an openidConnectClient configuration in a non-deterministic manner is the _current_ behavior since the iteration order may vary depending on the JDK and the OS. The design for this Epic takes that existing behavior into account and explicitly mentions it to leave no doubt that there is no specific order in which the configuration elements are read and processed.

As for 4, it is left to the chosen configuration to decide whether a 401 is done or a new authentication takes place since this is also the _current_ behavior controlled by the `inboundPropagation` attribute.

**Comment by teddyjtorres** [comment]
This feature's development is complete and it is just pending approvals for 22.0.0.9. The demo has been scheduled for the 22.15 EOI meeting. The PR for removing the beta guards is also ready in https://github.com/OpenLiberty/open-liberty/pull/21804.

**Comment by chirp1** [comment]
ID issue for this feature is at https://github.com/OpenLiberty/docs/issues/5646 .

**Comment by teddyjtorres** [comment]
Targeting to 220010 since Epic is still missing FAT, Demo, and Globalization approvals.

**Comment by teddyjtorres** [comment]
Moved demo to next EOI meeting.

**Comment by teddyjtorres** [comment]
@OpenLiberty/globalization-approvers - There are no new messages. Please let me know if anything else is needed. Thanks.

**Comment by teddyjtorres** [comment]
@OpenLiberty/demo-approvers - Please let me know if I can get an exemption from doing a demo for this small improvement. We're currently working on other higher priority items.

**Comment by tevans78** [comment]
> @OpenLiberty/demo-approvers - Please let me know if I can get an exemption from doing a demo for this small improvement. We're currently working on other higher priority items.

Demo approval given because the demo has been scheduled. You may yet postpone the demo to a later EOI if that is more convenient.

**Comment by teddyjtorres** [comment]
All approvals for this "No Design" feature are complete. Merged https://github.com/OpenLiberty/open-liberty/pull/21804 to remove beta guards.

**Comment by teddyjtorres** [comment]
Closing this Epic. Commit is found in latest green release build, Build Liberty Official Release Build - Static cl221020220829-0300.
